frdrpc/rpcserver.go

```diff
@@ -148,10 +156,28 @@ func (s *RPCServer) Start() error {
 		return errServerAlreadyStarted
 	}
 
+	// Start the macaroon service and let it create its default macaroon in
+	// case it doesn't exist yet.
+	if err := s.startMacaroonService(); err != nil {
+		return fmt.Errorf("error starting macaroon service: %v", err)
+	}
+
+	// First we add the security interceptor to our gRPC server options that
+	// checks the macaroons for validity.
+	serverOpts := s.macaroonInterceptor()
+	s.grpcServer = grpc.NewServer(serverOpts...)
+
 	// Start the gRPC RPCServer listening for HTTP/2 connections.
 	log.Info("Starting gRPC listener")
 	grpcListener, err := net.Listen("tcp", s.cfg.RPCListen)
 	if err != nil {
+		// The macaroon service is the only thing we started yet, so
+		// clean that up now before we exit.
+		if err := s.stopMacaroonService(); err != nil {
```

Comment:
Not sure whether this is more simple, but we could have:
```
shutdownFuncs []func()error
defer func(){
  s:= range shutdownFuncs; s()
}
```

Then just append as we go and set the slice to nil if we startup successfully.

Reply:
I like that idea, done.